### DISCUSSION TITLE
Remove pyarrow requirement for Polars import/export

### DISCUSSION THREAD

**Discussion by kylebarron** [duckdb/duckdb#13827] [Ideas]
Now that support for the [Arrow PyCapsule Interface](https://arrow.apache.org/docs/format/CDataInterface/PyCapsuleInterface.html) has been implemented in https://github.com/duckdb/duckdb/discussions/10716, it's possible to exchange Arrow data with Polars without pyarrow installed. However, this doesn't work _in practice_ (as of DuckDB 1.1) because DuckDB checks if the input is a Polars object **before** checking for existence of the Arrow PyCapsule Interface. So reading Polars data works without pyarrow only if you hide the fact that it's a `polars.DataFrame`.

This works:

```py
import duckdb
import polars as pl


class ArrowStream:
    def __init__(self, obj):
        self.obj = obj

    def __arrow_c_stream__(self, requested_schema=None):
        return self.obj.__arrow_c_stream__(requested_schema=requested_schema)


df = pl.DataFrame({"a": [1, 2, 3, 4]})
stream = ArrowStream(df)
con = duckdb.connect()
# Reading from the wrapper
sql = "SELECT * FROM stream"
con.query(sql)
```

```
┌───────┐
│   a   │
│ int64 │
├───────┤
│     1 │
│     2 │
│     3 │
│     4 │
└───────┘
```

This fails:

```py
import duckdb
import polars as pl

df = pl.DataFrame({"a": [1, 2, 3, 4]})
con = duckdb.connect()
# Reading from the DF directly
sql = "SELECT * FROM df"
con.query(sql)
```

```
---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
<ipython-input-6-3b52d8a1ce20> in ?()
----> 1 con.query(sql)

~/tmp/duckdb-1.1-pycapsule/.venv/lib/python3.11/site-packages/polars/_utils/deprecation.py in ?(*args, **kwargs)
     87         def wrapper(*args: P.args, **kwargs: P.kwargs) -> T:
     88             _rename_keyword_argument(
     89                 old_name, new_name, kwargs, function.__qualname__, version
     90             )
---> 91             return function(*args, **kwargs)

~/tmp/duckdb-1.1-pycapsule/.venv/lib/python3.11/site-packages/polars/dataframe/frame.py in ?(self, compat_level)
   1591             compat_level = False  # type: ignore[assignment]
   1592         elif isinstance(compat_level, CompatLevel):
   1593             compat_level = compat_level._version  # type: ignore[attr-defined]
   1594 
-> 1595         record_batches = self._df.to_arrow(compat_level)
   1596         return pa.Table.from_batches(record_batches)

ModuleNotFoundError: No module named 'pyarrow'
```

This should be simple to fix: just check for `__arrow_c_stream__` _before_ checking `isinstance(input, pl.DataFrame)`. This supports polars DataFrames from before it added support for the PyCapsule Interface (in 1.3.0: https://github.com/pola-rs/polars/pull/17676)

Versions are:

- DuckDB: 1.1.0
- Polars: 1.6.0

**Comment by Tishj** [comment]
We noticed shortly before release of 1.1 that there was a behavioral regression in using `__arrow_c_stream__` for Polars, so we opted to use the old way of scanning until that is resolved

Context:<https://github.com/duckdb/duckdb/pull/13798>

**Reply by kylebarron** [reply]
Oh right... I already forgot about that and https://github.com/duckdb/duckdb/issues/13793 😅. 

I was discussing this situation with @jorisvandenbossche, and separately the question of "when can you read a stream multiple times" has surfaced multiple times in the implementation of this protocol.

In general, it's not possible to reliably call `__arrow_c_stream__` multiple times. And so I suppose your hardship is that you don't know whether DuckDB should copy or view the stream's data. The nice part of knowing that the input object is a `polars.DataFrame` is knowing that you can reference the data multiple times. 

(This points to _one possible way_ to resolve the specific matter at hand: marking some known objects like `polars.DataFrame` as C Stream objects that can be queried multiple times, which still gets the benefit of not needing pyarrow)

I did see that that in the case of a polars `LazyFrame`, DuckDB will [materialize that stream](https://github.com/duckdb/duckdb/blob/fa5c2fe15f3da5f32397b009196c0895fce60820/tools/pythonpkg/src/python_replacement_scan.cpp#L112-L113), but it makes sense that DuckDB doesn't want to make a copy of all Arrow input.

**Reply by arambert** [reply]
Hi,
First of all thank you for this great library!!

We're using polars and duckdb in a bundled python application and for now we need to install pyarrow. Pyarrow alone adds around **100MB** in the package... We have no other usage for it.

Is it still on your roadmap to make this dependency unrequired for polars Dataframes to be readable into duckdb?
Do you know any way to get rid of it? (like in the first example of the initial post of the discussion for instance)

**Reply by Tishj** [reply]
When the object is of a class we recognize as polars/arrow we require pyarrow, you could use the `__arrow_c_stream__` method of the polars dataframe to create a capsule, then do the scan on the capsule instead of the polars dataframe

That should circumvent the problem

**Reply by arambert** [reply]
Thanks!
For future reference, this is how I removed the need of having pyarrow in my dependencies: 

## To be able to read Polars dataframes in duckdb without pyarrow

```python
# define a class implementing the __arrow_c_stream__ method on the object for duckdb to be able to read it
class ArrowStream:
    def __init__(self, obj):
        self.obj = obj

    def __arrow_c_stream__(self, requested_schema=None):
        return self.obj.__arrow_c_stream__(requested_schema=requested_schema)

# then if polars_dataframe is a Polars DataFrame: 
df = ArrowStream(polars_dataframe)
duckdb.sql("SELECT * FROM df") # it works fine now

# a couple of times (with updates for instance) I've seen an error like "INTERNAL Error: Attempted to access index N within vector of size N", I've fixed it by wrapping the dataframe in a SELECT clause:
duckdb.sql("UPDATE tablename SET col1=df.col1 FROM (SELECT * FROM df) as df") # it works fine now
```

## To be able to export duckdb relations to Polars dataframes without pyarrow
```python
import polars as pl

# duckdb.sql("SELECT * FROM tablename").pl() doesn't work without pyarrow
# luckily polars knows how to read C arrow streams
# so do this instead:
pl.DataFrame(duckdb.sql("SELECT * FROM tablename"))
```

**Reply by henryharbeck** [reply]
Is this able to be picked up?

DuckDB certainly seems willing to integrate nicely with Polars (e.g., the deeper integration here https://github.com/duckdb/duckdb/pull/17947) - I am very appreciative of this.

Regarding 
>  the question of "when can you read a stream multiple times" has surfaced multiple times in the implementation of this protocol.

Does https://github.com/duckdb/duckdb/pull/17087 fix this (or at least get around it)?
And does that mean this can now be completed without causing https://github.com/duckdb/duckdb/issues/13793 regression again?

**Reply by henryharbeck** [reply]
Any chance to do this? It seems like blocking issues and concerns have been resolved.
It seems like a shame for the PyCapsule interface to be implemented, but not utilised on the DuckDB side for Polars specifically.

> When the object is of a class we recognize as polars/arrow we require pyarrow, you could use the __arrow_c_stream__ method of the polars dataframe to create a capsule, then do the scan on the capsule instead of the polars dataframe

This does indeed work, but unfortunately requires end users to know about `__arrow_c_stream__ ` (perhaps not likely).

--- 
FWIW, some slightly simpler workarounds than those above when going between DuckDB relations.

```python
import duckdb
import polars as pl
rel = duckdb.sql("SELECT 1 AS a")
df = pl.DataFrame({"b": 2})

# DuckDB to Polars

# rel.pl() # ModuleNotFoundError: No module named 'pyarrow'

pl.DataFrame(rel)
# shape: (1, 1)
# ┌─────┐
# │ a   │
# │ --- │
# │ i32 │
# ╞═════╡
# │ 1   │
# └─────┘

# Polars to DuckDB

# duckdb.sql("FROM df") # ModuleNotFoundError: No module named 'pyarrow'

cap = df.__arrow_c_stream__()
duckdb.sql("FROM cap")
# ┌───────┐
# │   b   │
# │ int64 │
# ├───────┤
# │     2 │
# └───────┘
```

**Reply by cmdlineluser** [reply]
@henryharbeck It seems the Python client has its own repo now https://github.com/duckdb/duckdb-python

Perhaps creating a new discussion/issue there is the way forward?

**Reply by henryharbeck** [reply]
Good call. Created https://github.com/duckdb/duckdb-python/discussions/132
